Button elements are clipping overflow by default

Ahmad-S792 opened the pull request:
#### 7db4da203517ad5d8bfdc28593054beedbad2a74
<pre>
Button elements are clipping overflow by default

Button elements are clipping overflow by default
<a href="https://bugs.webkit.org/show_bug.cgi?id=173872">https://bugs.webkit.org/show_bug.cgi?id=173872</a>

Reviewed by Aditya Keerthi.

This patch is to align Webkit with Gecko / Firefox and Blink / Chromium.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=193213

By default, the button element has clipping overflow and it is not aligned with web
and causes webcompat issues (e.g., <a href="http://nydailynews.com">http://nydailynews.com</a> (see the feedback button)).

* Source/WebCore/rendering/RenderButton.cpp:
(RenderButton::hasControlClip): Add new function
* Source/WebCore/rendering/RenderButton.h: Update &quot;hasControlClip&quot; to not return
true by default
* LayoutTests/fast/forms/control-clip.html: Rebaselined
* LayoutTests/fast/forms/control-clip-overflow.html: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-overflow-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-overflow-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/control-clip-overflow-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-overflow-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-overflow-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/257439@main">https://commits.webkit.org/257439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a1d2cf9c5e6c66d4fa9394c9024a6ad8e885b52

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98917 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108329 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168584 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102872 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8679 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91447 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106302 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104610 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90209 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33605 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88399 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21554 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76456 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2032 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1939 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5123 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6897 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42544 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3344 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderButton.cpp:
```diff
@@ -169,4 +170,10 @@ void RenderButton::layout()
 }
 #endif
 
+// For compatibility with IE/FF we only clip overflow on input elements.
```
I would rewrite this as `// Only clip overflow on input elements, to match other browsers.`, since this is also compatible with Blink.

Ahmad-S792 replied:
Done!

pxlcoder commented on Source/WebCore/rendering/RenderButton.cpp:
```diff
@@ -169,4 +170,10 @@ void RenderButton::layout()
+// For compatibility with IE/FF we only clip overflow on input elements.
+bool RenderButton::hasControlClip() const
+{
+    return !is<HTMLButtonElement>(formControlElement());
```
I think we should use `is<HTMLInputElement>` to avoid unnecessary negation, and better align with the comment.

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
Button elements are clipping overflow by default

Button elements are clipping overflow by default
https://bugs.webkit.org/show_bug.cgi?id=173872

Reviewed by Aditya Keerthi.

This patch is to align Webkit with Gecko / Firefox and Blink / Chromium.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=193213

By default, the button element has clipping overflow and it is not aligned with web
and causes webcompat issues (e.g., http://nydailynews.com (see the feedback button)).

* Source/WebCore/rendering/RenderButton.cpp:
(RenderButton::hasControlClip): Add new function
* Source/WebCore/rendering/RenderButton.h: Update "hasControlClip" to not return
true by default
* LayoutTests/fast/forms/control-clip.html: Rebaselined
* LayoutTests/fast/forms/control-clip-overflow.html: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-overflow-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/forms/control-clip-overflow-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/control-clip-overflow-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-overflow-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/forms/control-clip-overflow-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/257439@main
